tensorflow/core/grappler/optimizers/auto_mixed_precision.cc

```diff
@@ -1970,10 +1989,18 @@ int GetNumGPUs(const Cluster& cluster,
   int num_gpus = 0;
   for (const auto& device : devices) {
     const DeviceProperties& device_properties = device.second;
+#if GOOGLE_CUDA
```

Comment:
This seems to duplicate logic from `IsOnSuitableGPUArch`?

Reply:
This is counting the number of suitable GPUs in the cluster, in this regard I replicated the CUDA implementation using the rocm functionality. Here we are looping through the devices on a cluster and operating directly using their device properties, whereas `IsOnSuitableGPUArch` operates using nodes in the graph.

Reply:
Then could we factor out the code which returns a bool from the device and call that from both?

Reply:
I refactored `HasFastFP16Support` to do both GPU checks. This way the `#if`s are only in that function, cleans up both `IsOnSuitableGPU` and `GetNumGPUs`.